libraries/rush-lib/src/logic/operations/AsyncOperationQueue.ts

```diff
@@ -49,27 +56,67 @@ export class AsyncOperationQueue
     return promise;
   }
 
+  /**
+   * Set a callback to be invoked when one operation is completed.
+   * If all operations are completed, set the queue to done, resolve all pending iterators in next cycle.
+   */
+  public complete(): void {
+    this._completedOperations++;
+    if (this._completedOperations === this._totalOperations) {
+      this._isDone = true;
+    }
+  }
+
   /**
    * Routes ready operations with 0 dependencies to waiting iterators. Normally invoked as part of `next()`, but
    * if the caller does not update operation dependencies prior to calling `next()`, may need to be invoked manually.
    */
   public assignOperations(): void {
     const { _queue: queue, _pendingIterators: waitingIterators } = this;
 
+    if (this._isDone) {
+      for (const resolveAsyncIterator of waitingIterators.splice(0)) {
+        resolveAsyncIterator({
+          value: undefined,
+          done: true
+        });
+      }
+      return;
+    }
+
     // By iterating in reverse order we do less array shuffling when removing operations
     for (let i: number = queue.length - 1; waitingIterators.length > 0 && i >= 0; i--) {
       const operation: OperationExecutionRecord = queue[i];
 
-      if (operation.status === OperationStatus.Blocked) {
+      if (
```

Comment:
> the main change to this file is that operations are left in the queue during execution, and only removed once the operation has a result?

Yes. you are right. That comes from cobuild tasks may be marked as "RemotePending" and check it later.


> Worth noting that this will cause the time to obtain the next operation to scale linearly with the parallelism of the execution, since it will tend to have to scan through however many operations are currently executing to find the next available operation on every call.

Nice catch. I thought this penalty before. A simple idea is

Deciding whether there will be a "remotePending" and pass it to the "AsyncQueue". In this way. "AsyncQueue" can have a code path which make it back to original behavior in non-cobuild scenario. The drawback is that it will make "AsyncQueue" hard to maintain...